lib/services/security/users.js

```diff
@@ -334,9 +356,14 @@ function getUserNoGroups(userName, callback) {
     if (!found) return callback(null, null);
     const password = found.data.password;
     const returnUser = this.securityService.serialize('user', found);
-    this.__cache_users.setSync(preparedUserName, returnUser, { clone: false });
-    this.__cache_passwords.setSync(preparedUserName, password, { clone: false });
-    callback(null, returnUser);
+    this.permissionManager
+      .attachPermissions(returnUser)
+      .then(attached => {
+        this.__cache_users.setSync(preparedUserName, attached, { clone: false });
+        this.__cache_passwords.setSync(preparedUserName, password, { clone: false });
+        callback(null, attached);
```

Comment:
Note that this pattern risks calling callback twice. If the code executed inside the .then() callback throws an exception, you'll find yourself calling it again in the catch() block. This can cause very strange complications.

Reply:
the only way I can see this being a problem, is if the callback itself throws an error, in which case we should just wrap a try catch around it. is this what you're referring to?

Reply:
I would suggest the following:
1. make a new async function (perhaps called _getUserNoGroups) using pure async/await (and promisifying any callback-based functions it needs to use)
2. restore the callback-driven getUserNoGroups using util.promisify:
```js
const getUserNoGroups = util.callbackify(_getUserNoGroups);
```

Reply:
what I've done is:
1: convert getUserNoGroups into an async/await function
2: used util.callbackify() when adding the method to the *SecurityUsers.prototype*

is this what you meant?